app/models/voucher.rb

```diff
@@ -1,15 +1,136 @@
+    return if order.vouchers.empty?
+
+    # We only support one voucher per order right now, we could just loop on vouchers
+    adjustment = order.vouchers.first
+
+    # Recalculate value
+    amount = adjustment.originator.compute_amount(order)
+
+    if order.additional_tax_total.positive?
+      handle_tax_excluded_from_price(order, amount)
+    else
+      handle_tax_included_in_price(order, amount)
+    end
+
+    # Move to closed state
+    adjustment.close
+  end
+
+  def self.handle_tax_excluded_from_price(order, amount)
+    voucher_rate = amount / order.total
+
+    # TODO: might need to use VoucherTax has originator (sub class of Voucher)
+    # Adding the voucher tax part
+    tax_amount = voucher_rate * order.additional_tax_total
+
+    adjustment = order.vouchers.first
+    adjustment_attributes = {
+      amount: tax_amount,
+      originator: adjustment.originator,
+      order: order,
+      label: "Tax #{adjustment.label}",
+      mandatory: false,
+      state: 'closed',
+      tax_category: nil,
+      included_tax: 0
+    }
+    order.adjustments.create(adjustment_attributes)
+
+    # Update the adjustment amount
+    amount = voucher_rate * (order.total - order.additional_tax_total)
+
+    adjustment.update_columns(
+      amount: amount,
+      updated_at: Time.zone.now
+    )
+  end
+
+  def self.handle_tax_included_in_price(order, amount)
+    voucher_rate = amount / order.total
+    included_tax = voucher_rate * order.included_tax_total
+
+    # Update Adjustment
+    adjustment = order.vouchers.first
+
+    return unless amount != adjustment.amount || included_tax != 0
+
+    adjustment.update_columns(
+      amount: amount,
+      included_tax: included_tax,
+      updated_at: Time.zone.now
+    )
+  end
+
   def value
     10
   end
 
   def display_value
     Spree::Money.new(value)
   end
+
+  # override the one from CalculatedAdjustments
+  # Create an "open" adjustment which will be updated later once tax and other fees have
+  # been applied to the order
+  #
+  # rubocop:disable Style/OptionalBooleanParameter
+  def create_adjustment(label, order, mandatory = false, _state = "open", tax_category = nil)
+    amount = compute_amount(order)
+
+    adjustment_attributes = {
+      amount: amount,
+      originator: self,
+      order: order,
+      label: label,
+      mandatory: mandatory,
+      state: "open",
+      tax_category: tax_category
+    }
+
+    order.adjustments.create(adjustment_attributes)
+  end
+  # rubocop:enable Style/OptionalBooleanParameter
+
+  # override the one from CalculatedAdjustments so we limit adjustment to the maximum amount
+  # needed to cover the order, ie if the voucher covers more than the order.total we only need
+  # to create an adjustment covering the order.total
+  # Doesn't work with taxes for now
+  # TODO move this to a calculator
+  def compute_amount(order)
+    amount = calculator.compute(order)
+
+    return -order.total if amount.abs > order.total
+
+    amount
```

Comment:
What do you think about this one?
```suggestion
    calculator.compute(order).clamp(-order.total, 0)
```

Reply:
TIL about `clamp` :tada: